Force success_key value against play_context and connection play_context

Yannig opened the pull request:
Do not look for sudo prompt on second item run (using with_items for exemple) for the same task.

Fix https://github.com/ansible/ansible/issues/13624

Yannig commented:
The tests are broken due to a bug in make tests-py3 (https://github.com/ansible/ansible/issues/13638)

jimi-c commented:
@Yannig a couple of things:

1) I don't think the flag value should be a class variable, but should be an instance variable instead.
2) It looks like you've brought in one of @abadger's patches, which fixes the py3 bug, but PRs shouldn't include patches from others, so this needs to be rebased on that.

Yannig commented:
@jimi-c done and happy new years from France!

jimi-c commented:
Looks good to me, cc'ing @amenonsen as he wrote that code originally.

jimi-c commented:
And Happy New Years to you too!

amenonsen commented:
@jimi-c no, I didn't touch local.py, only ssh.py. (But happy new year to all of you anyway! :-)

jimi-c commented:
Doh, thought this was about ssh.py, wasn't paying attention...

bcoca commented:
this should not be needed as `self.check_become_success(become_output)` should see the `echo successkey` string and not wait for the sudo prompt

Yannig commented:
@bcoca Agree but if you take a look at https://github.com/ansible/ansible/issues/13624, you'll see that - with items or something similar - Ansible crash saying that he fail to become the become_user (**ERROR! privilege output closed while waiting for password prompt:**).

If you launch the same playbook without sudo password (-K options), Ansible succeed in launching the playbook. The PR here fix this particular issue.  Maybe there's a better way. Maybe a bug in `self.check_become_success()` need to be fixed ...

Yannig commented:
Here is the playbook I'm using for doing my tests: https://github.com/Yannig/ansible-issue-low-speed/blob/master/local-connector/local-connector.yml

And here the command line I'm using:

`ansible-playbook local-connector/local-connector.yml -K`

Yannig commented:
Digging into play_context, I saw 2 differents keys generated for the sudo prompt (success_key) but Ansible is mixing up the keys and expect the wrong success_key.

Yannig commented:
@jimi-c this time, no more boolean, just using the same condition as in ssh.py to call check_become_success.

Yannig commented:
@bcoca what do you think about this last version?

Yannig committed:
Force success_key value between play_context and connection play_context.
Fix https://github.com/ansible/ansible/issues/13728 and https://github.com/ansible/ansible/issues/13624

Yannig committed:
Check if success_key is present in each line in output string. Handle cases where output buffer is compound of multiple lines.
Fix for https://github.com/ansible/ansible/issues/13763

Yannig commented:
@amenonsen The last commit may fix the remaining issue.

I was wondering if you are also running Ansible on a MacOS.

bcoca commented:
@Yannig if this fixes it for you, I think we have a deeper issue with the forks giving out incorrect play_context to the plugins that should be in lockstep.

ktaragorn commented:
@bcoca the problem is that the `connection` and with that the `play_context` is reused on subsequent runs, that is why it is out of sync

Yannig commented:
@bcoca @ktaragorn looking after this problem, Ansible have the following behaviour:
- create new connection + play_context object each time when using ssh + with_items
- create only one connection but multiple play_context when using local connector + with_items

Well, I don't know which one is wrong and I'm wondering if it maybe a better idea to have the same behaviour for every connection objects.

Yannig commented:
@bcoca I have a different way to get this issue fixed:

``` patch
diff --git a/lib/ansible/executor/task_executor.py b/lib/ansible/executor/task_executor.py
index 4a2d30a..e59860e 100644
--- a/lib/ansible/executor/task_executor.py
+++ b/lib/ansible/executor/task_executor.py
@@ -517,7 +517,7 @@ class TaskExecutor:
             'normal',
             task=async_task,
             connection=self._connection,
-            play_context=self._play_context,
+            play_context=self._connection._play_context,
             loader=self._loader,
             templar=templar,
             shared_loader_obj=self._shared_loader_obj,
@@ -590,7 +590,7 @@ class TaskExecutor:
                 'normal',
                 task=self._task,
                 connection=ssh_connection,
-                play_context=self._play_context,
+                play_context=ssh_connection._play_context,
                 loader=self._loader,
                 templar=templar,
                 shared_loader_obj=self._shared_loader_obj,
@@ -635,7 +635,7 @@ class TaskExecutor:
             handler_name,
             task=self._task,
             connection=connection,
-            play_context=self._play_context,
+            play_context=connection._play_context,
             loader=self._loader,
             templar=templar,
             shared_loader_obj=self._shared_loader_obj,
```

Basicaly, we use the same play context object as in connection object.

It fix local connector and do not break other things.

ktaragorn commented:
I think we need @cchurch  to weigh in on his motivations on making that change, and on the right way to fix this issue. The `_play_context` always used go be refreshed every iteration before that commit, so the change you suggest could look like a bug in that context @Yannig, depends on the actual utility of making it change or not.
